docs/design.md

```diff
@@ -967,6 +864,92 @@ spare capacity is chosen in the same datacenter and a special-case split
+elections. Cockroach weights random timeouts such that the replicas with
+shorter round trip times to peers are more likely to hold elections
+first (not implemented yet). Only the Raft leader may propose commands;
+followers will simply relay commands to the last known leader.
+
+Our Raft implementation was developed together with CoreOS, but adds an extra
+layer of optimization to account for the fact that a single Node may have
+millions of consensus groups (one for each Range). Areas of optimization
+are chiefly coalesced heartbeats (so that the number of nodes dictates the
+number of heartbeats as opposed to the much larger number of ranges) and
+batch processing of requests.
+Future optimizations may include two-phase elections and quiescent ranges
+(i.e. stopping traffic completely for inactive ranges).
+
+# Range Leadership (Leader Leases)
+
+As outlined in the Raft section, the replicas of a Range are organized as
+a Raft group and execute commands from their shared commit log. Going
+through Raft is an expensive operation though, and there are tasks which
+should only be carried out by a single replica at a time (as opposed to
+all of them).
+For these reasons, Cockroach introduces the concept of a 
+**Range Leader**: This is a lease held for a slice of (database, i.e. hybrid
+logical) time and is established by committing a special log entry through Raft
+containing the interval the leadership is going to be active on, along with the
+Node:RaftID combination that uniquely describes the requesting replica.
+Reads and writes must generally be addressed to the replica holding the lease;
+if none does, any replica may be addressed, causing it to try to obtain the
+lease synchronously. Requests received by a non-leader (for the HLC timestamp
+specified in the request's header) fail with an error pointing at the replica's
+last known leader. These requests are retried transparently with the updated
+leader by the gateway node and never reach the client.
+
+The replica holding the lease is in charge of handling Range-specific
+maintenance tasks such as
+
+* gossiping the sentinel and/or first range information
+* splitting, merging and rebalancing
+* handling message queues (see below)
+
+and, very importantly, may satisfy reads locally, without incurring the
+overhead of going through Raft.
+
+Since reads bypass Raft, a new lease holder will, among other things, ascertain
+that its timestamp cache does not report timestamps smaller than the previous
+lease holder's (so that it's compatible with reads which may have occurred on
+the former leader). This is accomplished by setting the low water mark of the
+timestamp cache to the expiration of the previous lease plus the maximum clock
+offset.
+
+## Relationship to Raft leadership
+
+Range leadership is completely separate from Raft leadership. It's normal and
+expected that in practice, the Raft leader and the Range leader may not be the
```

Comment:
I can't parse the sentence with your update. Instead I rewrote (simplified it), PTAL.